### ISSUE TITLE
Is DiGraph.predecessors() slow?

### ISSUE THREAD

**Issue by alabarre** [networkx/networkx#3448]
While working on [an improved algorithm for transitive closure for DAGs](https://github.com/networkx/networkx/pull/3445) I noted that what was dragging my algorithm down seemed to be the calls to `DiGraph.predecessors`. The two functions return exactly the same result as `transitive_closure`:

~~~python3
def transitive_closure_nx_better_dag(G):
    TC = G.copy()

    for vertex in list(nx.algorithms.dag.topological_sort(G))[-1::-1]:
        TC.add_edges_from((vertex, u) for u in descendants_at_distance(TC, vertex, 2))    # defined below

    return TC

def transitive_closure_nx_better_dag_pred(G):
    TC = G.copy()
    
    for vertex in list(nx.algorithms.dag.topological_sort(G))[-1::-1]:
        TC.add_edges_from(product(TC.predecessors(vertex), TC.successors(vertex)))
    
    return TC
~~~

But the timings are very different as you can see below, which is one of the reasons why I added the function:

~~~python3
def descendants_at_distance(G, source, distance):
    """Returns descendants at a given distance from a source. 
    Similar to bfs_successors except we don't keep the vertices closer than distance.    
    """
    current_distance = 0
    queue = deque([source])
    visited = {source}
    while queue:
        if current_distance == distance:
            return set(queue)
        current_distance += 1
        
        newqueue = deque()
        for vertex in queue:
            for child in G.successors(vertex):
                if child not in visited:
                    visited.add(child)
                    newqueue.appendleft(child)
                    
        queue = newqueue
        
    return set()
~~~

Here are my results for a single run on a random DAG (checks are basically asserts comparing the edge and vertex sets):

    $ python3 better_transitive_closure_dag.py 
    Testing transitive closure on a random DAG with 800 nodes and 239513 edges
    Running time for NetworkX's TC+:                         16.93851137161255
    Running time for transitive_closure_nx_better_dag:        3.103268623352051
    Running time for transitive_closure_nx_better_dag_pred:  64.09326243400574
    Checking that transitive_closure and transitive_closure_better agree ...
    ok
    Checking that transitive_closure and transitive_closure_better_pred agree ...
    ok

**Comment by dschult** [comment]
I'll have more time to look at this later, but just a first comment: In python, function calls are expensive relative to many other languages. So, ```G.predecessors(v)``` should be replaced almost everywhere performance matters by ```G.pred[v]```

**Comment by alabarre** [comment]
Good point, although at first glance it seems that the number of calls to `descendants_at_distance` in my other function will be the same as the number of calls to `predecessors` in the "bad" function. I tried your suggestion anyway, and while I was at it, I also replaced `TC.successors(v)` with `TC.succ[v]`; but performance doesn't seem to improve:

    $ python3 better_transitive_closure_dag.py 
    Testing transitive closure on a random DAG with 800 nodes and 239661 edges
    Running time for NetworkX's TC+:                        22.410506010055542
    Running time for transitive_closure_nx_better_dag:       4.086050510406494
    Running time for transitive_closure_nx_better_dag_pred: 81.44190216064453

I also tried replacing the call to `itertools.product` with a comprehension:

~~~python3
    TC.add_edges_from((u, v) for u in TC.pred[vertex] for v in TC.succ[vertex])
~~~

but results are similar.

**Comment by dschult** [comment]
Hmmm... Looks like there is more work being done when, starting from vertex, we go forward two steps than if we go forward one and backward one. Is ```better_dag_pred``` adding the same edges many more times?

I can't repeat your results. The DAG I use gives about the same running time for all 3 routines. The NX  and ```better_dag_pred``` are about 10% faster than ```better_dag```.  How are you creating your graphs?

**Comment by alabarre** [comment]
That is odd. Here's the function I wrote to generate random graphs:

**Comment by alabarre** [comment]
That is odd. Here's the function I wrote to generate random DAGs:

~~~python3
from random import random
import networkx as nx

def random_dag(number_of_nodes, edge_probability):
    G = nx.DiGraph()
    G.add_nodes_from(range(number_of_nodes))
    for i in range(number_of_nodes-1):
        for j in range(i+1, number_of_nodes):
            if random() < edge_probability:
                G.add_edge(i, j)
    return G
~~~

Your comment about `better_dag_pred` doing useless work like adding many existing edges is probably right, I don't see a better explanation for its (far) worse behaviour.

**Comment by dschult** [comment]
Are we done with this Issue?  An associated PR #3445 has been merged. But I'm not sure if this is a sepaarte issue or not.

**Comment by alabarre** [comment]
To me this is a separate issue. But you and I apparently get different results with the same code, and I don't know what your policy is in that kind of situation.

**Comment by dschult** [comment]
I have investigated further. Using your ```random_dag``` graph creator I get the same differences in speed that you report. (The DAGs I was using were very simple dags, so probably not good examples.) The two methods you use are vastly different in times, just as you report.

I then counted how many edges each method added to the graphs. Notice that the final graphs created are the same, so any difference in number of edges added is just the algorithm doing redundant work.

With 80 nodes, p=0.7 I got
```
Generated DAG with 80 nodes and 2201 edges
ahead 2 method: added 896 edges.
pred_succ method: added 55731 edges.
```
With 800 nodes p=0.7 I got
```
Generated DAG with 800 nodes and 223673 edges
ahead 2 method: added 95460 edges.
pred_succ method: added 59292027 edges.
```

So, I am going to close this Issue. While I don't have a small example to show that one algorithm is slower than the other, this evidence shows that for the examples we are looking at it creates fewer redundant edges to add if you look forward 2 descendants than if you look forward 1 and backward 1.
